File: apps/dashboard/src-tauri/src/state.rs
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::PathBuf;
use std::sync::{Arc, Mutex};

// --- Data types for API responses ---

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProjectState {
    pub name: String,
    pub path: String,
    #[serde(rename = "defaultBranch")]
    pub default_branch: String,
    pub worktrees: Vec<WorktreeState>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub label: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct WorktreeState {
    pub branch: String,
    pub path: String,
    pub head: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct AppState {
    pub projects: Vec<ProjectState>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct NotificationSettings {
    #[serde(
        rename = "soundOnNeedsAttention",
        default,
        skip_serializing_if = "Option::is_none"
    )]
    pub sound_on_needs_attention: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sound: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LabelDefinition {
    pub id: String,
    pub name: String,
    pub color: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct Settings {
    #[serde(rename = "worktreesDir", skip_serializing_if = "Option::is_none")]
    pub worktrees_dir: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub defaults: Option<serde_json::Value>,
    #[serde(rename = "codingAgent", skip_serializing_if = "Option::is_none")]
    pub coding_agent: Option<serde_json::Value>,
    #[serde(rename = "webServerPort", skip_serializing_if = "Option::is_none")]
    pub web_server_port: Option<u16>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub notifications: Option<NotificationSettings>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub labels: Option<Vec<LabelDefinition>>,
    #[serde(rename = "tokenSecret", skip_serializing_if = "Option::is_none")]
    pub token_secret: Option<String>,
    #[serde(rename = "tunnelSubdomain", skip_serializing_if = "Option::is_none")]
    pub tunnel_subdomain: Option<String>,
    #[serde(rename = "autoStartTunnel", skip_serializing_if = "Option::is_none")]
    pub auto_start_tunnel: Option<bool>,
}

// --- File system helpers (only for reading settings) ---

pub fn band_home() -> PathBuf {
    dirs::home_dir()
        .expect("Could not find home directory")
        .join(".band")
}

fn settings_file() -> PathBuf {
    band_home().join("settings.json")
}

pub fn load_settings() -> Result<Settings, String> {
    let path = settings_file();
    if !path.exists() {
        return Ok(Settings::default());
    }
    let data = fs::read_to_string(&path).map_err(|e| format!("Failed to read settings: {e}"))?;
    serde_json::from_str(&data).map_err(|e| format!("Failed to parse settings: {e}"))
}

// --- In-memory active workspace state ---

pub struct ActiveWorkspaceState(pub Arc<Mutex<Option<String>>>);

impl ActiveWorkspaceState {
    pub fn new() -> Self {
        Self(Arc::new(Mutex::new(None)))
    }
}

// --- Cached project state (refreshed from web server API) ---

pub struct CachedState {
    pub app_state: AppState,
}

#[derive(Clone)]
pub struct ProjectCache(Arc<Mutex<Option<CachedState>>>);

impl ProjectCache {
    pub fn new() -> Self {
        Self(Arc::new(Mutex::new(None)))
    }

    /// Get a copy of the current cached state, if available.
    pub fn get(&self) -> Option<AppState> {
        self.0.lock().ok()?.as_ref().map(|c| c.app_state.clone())
    }

    /// Update the cached state. Used by macOS focus polling (ide.rs).
    #[cfg_attr(not(target_os = "macos"), allow(dead_code))]
    pub fn set(&self, state: AppState) {
        if let Ok(mut guard) = self.0.lock() {
            *guard = Some(CachedState { app_state: state });
        }
    }
}

File: apps/web/src/lib/state.ts
import { randomBytes } from "node:crypto";
import { mkdirSync, readFileSync, renameSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";
import { dirname, join } from "node:path";
import { toWorkspaceId } from "@band-app/dashboard-core";
import { eq } from "drizzle-orm";
import { getDb } from "./db/connection";
import {
  branchStatuses as branchStatusesTable,
  projects as projectsTable,
  workspaceStatuses as workspaceStatusesTable,
  worktrees as worktreesTable,
} from "./db/schema";

export interface ProjectState {
  name: string;
  path: string;
  defaultBranch: string;
  worktrees: WorktreeState[];
  label?: string;
}

export interface WorktreeState {
  branch: string;
  path: string;
  head?: string;
}

export interface AppState {
  projects: ProjectState[];
}

export interface AgentInfo {
  name: string;
  status: string;
  lastActivity: string;
  summary?: string;
}

export interface WorkspaceStatus {
  workspaceId: string;
  project: string;
  branch: string;
  worktreePath: string;
  ide: string;
  agent?: AgentInfo;
}

export interface LabelDefinition {
  id: string;
  name: string;
  color: string;
}

export interface NotificationSettings {
  soundOnNeedsAttention?: boolean;
  sound?: string;
}

export interface Settings {
  worktreesDir?: string;
  defaults?: unknown;
  codingAgent?: {
    type: string;
    command?: string;
  };
  webServerPort?: number;
  notifications?: NotificationSettings;
  labels?: LabelDefinition[];
  tokenSecret?: string;
  autoStartTunnel?: boolean;
  /** Extra fields not explicitly modeled (e.g. Tauri app definitions). Preserved across read/write. */
  [key: string]: unknown;
}

export function bandHome(): string {
  if (process.env.BAND_HOME) return process.env.BAND_HOME;
  return join(homedir(), ".band");
}

export function loadState(): AppState {
  const db = getDb();
  const projectRows = db.select().from(projectsTable).orderBy(projectsTable.sortOrder).all();

  const worktreeRows = db.select().from(worktreesTable).all();

  const wtByProject = new Map<string, WorktreeState[]>();
  for (const row of worktreeRows) {
    const list = wtByProject.get(row.projectName) ?? [];
    list.push({
      branch: row.branch,
      path: row.path,
      head: row.head ?? undefined,
    });
    wtByProject.set(row.projectName, list);
  }

  return {
    projects: projectRows.map((row) => ({
      name: row.name,
      path: row.path,
      defaultBranch: row.defaultBranch,
      label: row.label ?? undefined,
      worktrees: wtByProject.get(row.name) ?? [],
    })),
  };
}

export function saveState(state: AppState): void {
  const db = getDb();

  db.transaction((tx) => {
    tx.delete(worktreesTable).run();
    tx.delete(projectsTable).run();

    for (let i = 0; i < state.projects.length; i++) {
      const project = state.projects[i];
      tx.insert(projectsTable)
        .values({
          name: project.name,
          path: project.path,
          defaultBranch: project.defaultBranch,
          label: project.label ?? null,
          sortOrder: i,
        })
        .run();

      for (const wt of project.worktrees) {
        tx.insert(worktreesTable)
          .values({
            projectName: project.name,
            branch: wt.branch,
            path: wt.path,
            head: wt.head ?? null,
          })
          .run();
      }
    }
  });
}

function settingsFile(): string {
  return join(bandHome(), "settings.json");
}

export function loadSettings(): Settings {
  try {
    const data = readFileSync(settingsFile(), "utf-8");
    return JSON.parse(data) as Settings;
  } catch {
    return {};
  }
}

export function saveSettings(settings: Settings): void {
  const filePath = settingsFile();
  // Merge with existing file contents to preserve unknown fields (e.g. Tauri extras)
  let existing: Record<string, unknown> = {};
  try {
    existing = JSON.parse(readFileSync(filePath, "utf-8"));
  } catch {
    // File doesn't exist or is invalid — start fresh
  }
  const merged = { ...existing, ...settings };
  const data = `${JSON.stringify(merged, null, 2)}\n`;
  // Atomic write: write to temp file then rename
  const dir = dirname(filePath);
  mkdirSync(dir, { recursive: true });
  const tmpPath = `${filePath}.tmp.${process.pid}`;
  writeFileSync(tmpPath, data, "utf-8");
  renameSync(tmpPath, filePath);
}

export function getOrCreateToken(): string {
  const settings = loadSettings();
  if (settings.tokenSecret) return settings.tokenSecret;
  const token = randomBytes(32).toString("hex");
  const current = loadSettings();
  current.tokenSecret = token;
  saveSettings(current);
  return token;
}

export function worktreesDir(): string {
  const settings = loadSettings();
  return settings.worktreesDir ?? join(bandHome(), "worktrees");
}

export function loadCurrentStatuses(): WorkspaceStatus[] {
  const db = getDb();
  const rows = db.select().from(workspaceStatusesTable).all();
  return rows.map((row) => ({
    workspaceId: row.workspaceId,
    project: row.project,
    branch: row.branch,
    worktreePath: row.worktreePath,
    ide: row.ide,
    agent: row.agentName
      ? {
          name: row.agentName,
          status: row.agentStatus ?? "unknown",
          lastActivity: row.agentLastActivity ?? "",
          summary: row.agentSummary ?? undefined,
        }
      : undefined,
  }));
}

export function getWorkspaceStatus(workspaceId: string): WorkspaceStatus | null {
  const db = getDb();
  const row = db
    .select()
    .from(workspaceStatusesTable)
    .where(eq(workspaceStatusesTable.workspaceId, workspaceId))
    .get();
  if (!row) return null;
  return {
    workspaceId: row.workspaceId,
    project: row.project,
    branch: row.branch,
    worktreePath: row.worktreePath,
    ide: row.ide,
    agent: row.agentName
      ? {
          name: row.agentName,
          status: row.agentStatus ?? "unknown",
          lastActivity: row.agentLastActivity ?? "",
          summary: row.agentSummary ?? undefined,
        }
      : undefined,
  };
}

export function upsertWorkspaceStatus(
  workspaceId: string,
  agent: { status: string; lastActivity?: string },
): WorkspaceStatus {
  const db = getDb();

  // Read existing row to preserve fields
  const existing = db
    .select()
    .from(workspaceStatusesTable)
    .where(eq(workspaceStatusesTable.workspaceId, workspaceId))
    .get();

  const now = Date.now();
  const mergedAgent = {
    agentName: existing?.agentName ?? "claude-code",
    agentStatus: agent.status,
    agentLastActivity: agent.lastActivity ?? existing?.agentLastActivity ?? "",
    agentSummary: existing?.agentSummary ?? null,
  };

  if (existing) {
    db.update(workspaceStatusesTable)
      .set({ ...mergedAgent, updatedAt: now })
      .where(eq(workspaceStatusesTable.workspaceId, workspaceId))
      .run();
  } else {
    // For new rows, resolve workspace identity from the worktrees DB
    const ws = resolveWorkspaceIdentity(workspaceId);
    db.insert(workspaceStatusesTable)
      .values({
        workspaceId,
        project: ws?.project ?? "",
        branch: ws?.branch ?? "",
        worktreePath: ws?.worktreePath ?? "",
        ide: "vscode",
        ...mergedAgent,
        updatedAt: now,
      })
      .run();
  }

  return getWorkspaceStatus(workspaceId)!;
}

function resolveWorkspaceIdentity(
  workspaceId: string,
): { project: string; branch: string; worktreePath: string } | null {
  const state = loadState();
  for (const proj of state.projects) {
    for (const wt of proj.worktrees) {
      if (toWorkspaceId(proj.name, wt.branch) === workspaceId) {
        return { project: proj.name, branch: wt.branch, worktreePath: wt.path };
      }
    }
  }
  return null;
}

export function deleteWorkspaceStatus(workspaceId: string): void {
  const db = getDb();
  db.delete(workspaceStatusesTable)
    .where(eq(workspaceStatusesTable.workspaceId, workspaceId))
    .run();
}

export function deleteBranchStatus(workspaceId: string): void {
  const db = getDb();
  db.delete(branchStatusesTable).where(eq(branchStatusesTable.workspaceId, workspaceId)).run();
}
